tests/test_warc_metadata_sidecar.py

```diff
@@ -166,22 +174,60 @@ def test_metadata_sidecar_image_record(self, mock_404, mock_language, mock_chara
                 if record.rec_type == 'metadata':
                     payload = record.content_stream().read()
         assert payload == img_payload
+        assert metadata_sidecar_return == (tmpdir / 'gif.warc.meta.gz', 1, 1)
 
     @patch('warc_metadata_sidecar.WARCWriter')
     def test_metadata_sidecar_revisit_record(self, mock_warcwriter, caplog, tmpdir):
         caplog.set_level(INFO)
-        sidecar.metadata_sidecar(str(tmpdir), REVISIT_TEST_FILE)
+        metadata_sidecar_return = sidecar.metadata_sidecar(str(tmpdir), REVISIT_TEST_FILE)
         assert 'Logging WARC metadata record information for %s', REVISIT_TEST_FILE in caplog.text
         assert 'Deleted sidecar, no records to collect.' in caplog.text
         assert tmpdir / 'revist.warc.meta.gz' not in tmpdir.listdir()
         mock_warcwriter.assert_called_once()
+        assert metadata_sidecar_return == (tmpdir / 'revisit.warc.meta.gz', 1, 0)
 
     def test_arc_record_has_no_concurrent_or_warcinfo_id(self, tmpdir):
-        sidecar.metadata_sidecar(str(tmpdir), ARC_TEST_FILE)
+        metadata_sidecar_return = sidecar.metadata_sidecar(str(tmpdir), ARC_TEST_FILE)
         path = os.path.join(tmpdir / 'text.warc.meta.gz')
         assert path in tmpdir.listdir()
+        assert metadata_sidecar_return == (tmpdir / 'text.warc.meta.gz', 2, 1)
         with open(path, 'rb') as stream:
             for record in ArchiveIterator(stream):
                 if record.rec_type == 'metadata':
                     assert record.rec_headers.get_header('WARC-Concurrent-ID') is None
                     assert record.rec_headers.get_header('WARC-Warcinfo-ID') is None
+
+    @patch('warc_metadata_sidecar.determine_soft404')
+    @patch('warc_metadata_sidecar.find_mime_and_puid')
+    @patch('warc_metadata_sidecar.find_character_set')
+    @patch('warc_metadata_sidecar.find_language')
+    @patch('warc_metadata_sidecar.create_string_payload', return_value='payload')
+    @patch('warc_metadata_sidecar.create_warcinfo_payload')
+    @patch('warc_metadata_sidecar.WARCWriter')
+    def test_digest_multiples_use_cache(self, mock_warcwriter, m_warcinfo, m_create_payload,
+                                        m_lang, m_charset, m_find_mime, m_soft404, caplog,
+                                        tmpdir):
+        # Clear the cache from previous tests
+        sidecar.DIGEST_CACHE = {}
+        # Get record digest from file to test DIGEST_CACHE
+        digest_list = []
+        with open(DIGEST_TEST_FILE, 'rb') as stream:
+            for record in ArchiveIterator(stream):
+                digest = record.rec_headers.get_header('WARC-Payload-Digest')
+                if digest and digest not in digest_list:
+                    digest_list.append(digest)
+        caplog.set_level(INFO)
+        writer = mock_warcwriter.return_value
+        m_find_mime.return_value = ({'python-magic': 'text/plain'}, None)
```

Comment:
I found it confusing to look at the WARC being tested that was 2 response records for gifs with matching digests and 2 response records with text with matching digests to then see that `m_lang.call_count` and `m_charset.call_count` were `2`. Looking at the test again, I see it is because of mocking the `m_find_mime` to always return text, even when a record is for a gif. I would expect from looking at the WARC that the  `m_lang.call_count` and `m_charset.call_count` would be `1` to show that there are two text files that are duplicates, meaning that with your changes using the digest to avoid multiple text processing calls, the language and charset aren't rechecked. Perhaps if `m_find_mime` had a `side_effect` for once being a gif and once being text instead of a `return_value` that always gives text, it would make the test more clear.

Reply:
Thank you! I was thinking that there was a way to give it multiple returns, but I was trying to do that with `return_value`. I forgot about side_effect. 🙁 I will fix that.